README.md
# Welcome to `cdk-elasticache-monitor`

ElasticacheAutoMonitor allows you to send email, sms, slack, or trigger aws lambda when an alarm occurs.
You will get the following monitoring:
  * Cpu Monitor: Should be less than 90%. (See below reference)
  * SwapUsage Monitor: Should be less than 50M.
  * Evictions Monitor: Should not have evictions value.
  * CurrConnections Monitor: According to your business needs, default every 1 vcup is equal to 200 connections.
  * FreeableMemory Monitor: Not less than 10%

Reference: [Here](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheMetrics.WhichShouldIMonitor.html)

## Usage
```typescript

const stack = new cdk.Stack();

// or sent with sms
ElasticacheAutoMonitor.setUpWithSms(stack, 'my-elasticache-id', {
  phones: ['+15533728278'],
  nodeTypeClass: NodeType.R4_16XLARGE,
});

// or sent with slack
ElasticacheAutoMonitor.setUpWithSlack(stack, 'my-elasticache-id', {
  webHookUrl: 'http://xxx.xxx.xxx',
  nodeTypeClass: NodeType.R4_16XLARGE,
});

// or trigger lambda
const fn = new lambda.Function(stack, 'MyFunction', {
  runtime: lambda.Runtime.NODEJS_10_X,
  handler: 'index.handler',
  code: lambda.Code.fromInline('exports.handler = function(event, ctx, cb) { return cb(null, "hi"); }'),
});
ElasticacheAutoMonitor.setUpWithLambda(stack, 'my-elasticache-id', fn, {
  nodeTypeClass: NodeType.R4_16XLARGE,
});
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.14](https://github.com/jialechan/cdk-elasticache-monitor/compare/v0.0.13...v0.0.14) (2020-08-28)


### Features

* **core:** ConstructLibraryAws API ([5320c36](https://github.com/jialechan/cdk-elasticache-monitor/commit/5320c366d3c62b61cca6a33bee8ce71c994ec2bc))

### 0.0.13 (2020-08-08)

### 0.0.12 (2020-08-08)

### [0.0.11](https://github.com/jialechan/cdk-elasticache-monitor/compare/v0.0.6...v0.0.11) (2020-08-08)

### 0.0.6 (2020-08-08)

### 0.0.5 (2020-08-08)

### 0.0.4 (2020-08-01)

### 0.0.3 (2020-08-01)

### 0.0.2 (2020-07-31)

### 0.0.1 (2020-07-31)


test/index.test.ts
import * as cdk from '@aws-cdk/core';
import '@aws-cdk/assert/jest';
import * as lambda from '@aws-cdk/aws-lambda';
import { ElasticacheAutoMonitor, NodeType } from '../src/index';

test('NodeType.T2_SMALL should equals 50 connections because T2_SMALL has 1 vcup', () => {
  const stack = new cdk.Stack();

  ElasticacheAutoMonitor.setUpWithSms(stack, 'my-elasticache-id', {
    phones: ['+123456789'],
    nodeType: NodeType.T2_SMALL,
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    MetricName: 'CurrConnections',
    Threshold: 200,
  });
});

test('NodeType.T2_MEDIUM should equals 100 connections because T2_MEDIUM has 2 cnup', () => {
  const stack = new cdk.Stack();

  ElasticacheAutoMonitor.setUpWithSms(stack, 'my-elasticache-id', {
    phones: ['+123456789'],
    nodeType: NodeType.T2_MEDIUM,
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    MetricName: 'CurrConnections',
    Threshold: 400,
  });
});

test('CPU should be less than 45% when vcpu are less than or eq 2', () => {
  const stack = new cdk.Stack();

  ElasticacheAutoMonitor.setUpWithSms(stack, 'my-elasticache-id', {
    phones: ['+123456789'],
    nodeType: NodeType.T2_MEDIUM,
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    MetricName: 'CPUUtilization',
    Threshold: 45,
  });
});

test('CPU should be less than 90% when vcpu are great than 2', () => {
  const stack = new cdk.Stack();

  ElasticacheAutoMonitor.setUpWithSms(stack, 'my-elasticache-id', {
    phones: ['+123456789'],
    nodeType: NodeType.R4_16XLARGE,
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    MetricName: 'EngineCPUUtilization',
    Threshold: 90,
  });
});

test('sent with sms', () => {
  const stack = new cdk.Stack();

  ElasticacheAutoMonitor.setUpWithSms(stack, 'my-elasticache-id', {
    phones: ['+123456789'],
    nodeType: NodeType.R4_16XLARGE,
  });

  expect(stack).toHaveResource('AWS::SNS::Subscription', {
    Protocol: 'sms',
  });
});

test('sent with slack', () => {
  const stack = new cdk.Stack();

  ElasticacheAutoMonitor.setUpWithSlack(stack, 'my-elasticache-id', {
    webHookUrl: 'http://xxx.xxx.xxx',
    nodeType: NodeType.R4_16XLARGE,
  });

  expect(stack).toHaveResource('AWS::SNS::Subscription', {
    Protocol: 'lambda',
  });
});

test('sent with lambda', () => {
  const stack = new cdk.Stack();

  const fn = new lambda.Function(stack, 'MyFunction', {
    runtime: lambda.Runtime.NODEJS_10_X,
    handler: 'index.handler',
    code: lambda.Code.fromInline('exports.handler = function(event, ctx, cb) { return cb(null, "hi"); }'),
  });

  ElasticacheAutoMonitor.setUpWithLambda(stack, 'my-elasticache-id', {
    lambda: fn,
    nodeType: NodeType.R4_16XLARGE,
  });

  expect(stack).toHaveResource('AWS::SNS::Subscription', {
    Protocol: 'lambda',
  });
});

test('Customize currConnections value', () => {
  const stack = new cdk.Stack();

  ElasticacheAutoMonitor.setUpWithSms(stack, 'my-elasticache-id', {
    phones: ['+123456789'],
    nodeType: NodeType.T2_SMALL,
    currConnectionsPeriod: cdk.Duration.minutes(2),
    currConnectionsThreshold: 200,
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    MetricName: 'CurrConnections',
    Period: 120,
    Threshold: 200,
  });
});

test('Default noteType', () => {
  const stack = new cdk.Stack();

  ElasticacheAutoMonitor.setUpWithSms(stack, 'my-elasticache-id', {
    phones: ['+123456789'],
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    MetricName: 'CurrConnections',
    Threshold: 400,
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    MetricName: 'CPUUtilization',
    Threshold: 45,
  });

  expect(stack).toHaveResource('AWS::CloudWatch::Alarm', {
    MetricName: 'FreeableMemory',
    Threshold: 685047284,
  });

});



.projenrc.js
const {
  ConstructLibraryAws
} = require('projen');

const AWS_CDK_LATEST_RELEASE = '1.61.0';
const PROJECT_NAME = 'cdk-elasticache-monitor';
const PROJECT_DESCRIPTION = 'ElasticacheAutoMonitor allows you to send email, sms, slack, or trigger aws lambda when an alarm occurs.';

const project = new ConstructLibraryAws({
  name: PROJECT_NAME,
  description: PROJECT_DESCRIPTION,
  authorName: "Jiale Chan",
  authorEmail: "jiale.chan@gmail.com",
  repository: "https://github.com/jialechan/cdk-elasticache-monitor.git",
  stability: 'experimental',

  keywords: [
    'cdk',
    'aws',
    'elasticache',
    'monitor',
    'alarm',
    'slack',
  ],

  catalog: {
    twitter: 'jialechan',
    announce: false,
  },

  // creates PRs for projen upgrades
  projenUpgradeSecret: 'PROJEN_GITHUB_TOKEN',
  cdkVersion: AWS_CDK_LATEST_RELEASE,
  cdkDependencies: [
    '@aws-cdk/core',
    '@aws-cdk/aws-lambda',
    '@aws-cdk/aws-cloudwatch',
    '@aws-cdk/aws-sns',
    '@aws-cdk/aws-sns-subscriptions',
    '@aws-cdk/aws-cloudwatch-actions',
  ],

  python: {
    distName: 'cdk-elasticache-monitor',
    module: 'cdk-elasticache-monitor'
  }
});

const common_exclude = ['cdk.out', 'cdk.context.json', 'images', 'yarn-error.log'];
project.npmignore.exclude(...common_exclude);
project.gitignore.exclude(...common_exclude);

project.synth();


src/cpu-monitor.ts
import * as cdk from '@aws-cdk/core';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import { NodeType } from './index'

// Reference: https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheMetrics.WhichShouldIMonitor.html#metrics-cpu-utilization
export class CpuMonitor {
  public static setup(scope: cdk.Construct, cacheClusterId: string, nodeType: NodeType) {

    let threshold = 90;
    let metricName = 'EngineCPUUtilization';

    if (nodeType && nodeType.vcupCount <= 2) {
      metricName = 'CPUUtilization';
      threshold = (90 / nodeType.vcupCount);
    }

    return new cloudwatch.Alarm(scope, 'alarm-elastc-cache-' + cacheClusterId + '-' + metricName, {
      alarmName: `ElasticCacheCPULoadAlarm[${cacheClusterId}]`,
      metric: new cloudwatch.Metric({
        namespace: 'AWS/ElastiCache',
        metricName,
        dimensions: { CacheClusterId: cacheClusterId },
        statistic: 'avg',
      }),
      threshold,
      period: cdk.Duration.minutes(1),
      evaluationPeriods: 1,
      alarmDescription: `The average CPU load(${metricName}) is greater than ${threshold}%.`,
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    });
  }
}



asset/lambda/index.py
import os
import json
import time
import urllib.parse
import urllib.request


def handler(event, context):
    """
    alarm to slack
    """

    print(json.dumps(event))

    slack_webhook_url = os.environ['SLACK_WEBHOOK_URL']
    channel = os.environ['CHANNEL']
    username = os.environ['USERNAME']
    icon_emoji = os.environ['ICON_EMOJI']

    for record in event.get("Records"):

        message = json.loads(record.get("Sns").get("Message"))

        title = message.get("AlarmName")
        info = message.get("AlarmDescription")
        newStateReason = message.get("NewStateReason")

        region = os.environ['AWS_REGION']

        log = "https://" + region + ".console.aws.amazon.com/cloudwatch/home?region=" + \
            region + "#alarmsV2:alarm/" + title + "?~(alarmStateFilter~'ALARM)"

        values = {
            "channel": channel,
            "username": username,
            "text": title + "\n" + info + "\n" + newStateReason + "\n" + "<" + log + "|AlarmState>",
            "icon_emoji": icon_emoji
        }

        params = json.dumps(values).encode('utf8')
        req = urllib.request.Request(slack_webhook_url, data=params, headers={
                                     'content-type': 'application/json'})
        response = urllib.request.urlopen(req)

        print(response.read())


src/evictions-monitor.ts
import * as cdk from '@aws-cdk/core';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';

// Reference: https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheMetrics.WhichShouldIMonitor.html#metrics-evictions
export class EvictionsMomiton {

  public static setup(scope: cdk.Construct, cacheClusterId: string) {

    return new cloudwatch.Alarm(scope, 'alarm-elastc-cache-' + cacheClusterId + '-Evictions', {
      alarmName: `ElasticCacheEvictionsAlarm[${cacheClusterId}]`,
      metric: new cloudwatch.Metric({
        namespace: 'AWS/ElastiCache',
        metricName: 'Evictions',
        dimensions: { CacheClusterId: cacheClusterId },
        statistic: 'max',
      }),
      threshold: 1,
      period: cdk.Duration.minutes(1),
      evaluationPeriods: 1,
      alarmDescription: 'Data evicted within one minute(We recommend that you determine your own alarm threshold for this metric based on your application needs)',
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    });
  }
}

src/swap-usage-monitor.ts
import * as cdk from '@aws-cdk/core';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';

// Reference: https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheMetrics.WhichShouldIMonitor.html#metrics-swap-usage
export class SwapUsageMomiton {
  public static setup(scope: cdk.Construct, cacheClusterId: string) {

    return new cloudwatch.Alarm(scope, 'alarm-elastc-cache-' + cacheClusterId + '-SwapUsage', {
      alarmName: `ElasticCacheSwapUsageAlarm[${cacheClusterId}]`,
      metric: new cloudwatch.Metric({
        namespace: 'AWS/ElastiCache',
        metricName: 'SwapUsage',
        dimensions: { CacheClusterId: cacheClusterId },
        statistic: 'max',
      }),
      threshold: 52430000,
      period: cdk.Duration.minutes(1),
      evaluationPeriods: 1,
      alarmDescription: 'The maximum value of the swapUsage within one minute is higher than 50MB.',
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    });
  }
}

src/freeable-memory-monitor.ts
import * as cdk from '@aws-cdk/core';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import { NodeType } from './index'

// According to business needs
export class FreeableMemoryMomiton {
  public static setup(scope: cdk.Construct, cacheClusterId: string, nodeType: NodeType) {

    const threshold = Math.round(nodeType.memory * 0.1);

    return new cloudwatch.Alarm(scope, 'alarm-elastc-cache-' + cacheClusterId + '-FreeableMemory', {
      alarmName: `ElasticCacheFreeableMemoryAlarm[${cacheClusterId}]`,
      metric: new cloudwatch.Metric({
        namespace: 'AWS/ElastiCache',
        metricName: 'FreeableMemory',
        dimensions: { CacheClusterId: cacheClusterId },
        statistic: 'min',
      }),
      threshold,
      period: cdk.Duration.minutes(1),
      evaluationPeriods: 1,
      alarmDescription: `The FreeableMemory value is less than ${(threshold / (1024 * 1024 * 1024)).toFixed(2)}Gb in one minute.`,
      comparisonOperator: cloudwatch.ComparisonOperator.LESS_THAN_THRESHOLD,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    });
  }
}

src/current-connections-monitor.ts
import * as cdk from '@aws-cdk/core';
import * as cloudwatch from '@aws-cdk/aws-cloudwatch';

// Reference: https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheMetrics.WhichShouldIMonitor.html#metrics-curr-connections
export class CurrConnectionsMomiton {
  public static setup(scope: cdk.Construct, cacheClusterId: string, period: cdk.Duration, threshold: number) {

    return new cloudwatch.Alarm(scope, 'alarm-elastc-cache-' + cacheClusterId + '-CurrConnections', {
      alarmName: `ElasticCacheCurrConnectionsAlarm[${cacheClusterId}]`,
      metric: new cloudwatch.Metric({
        namespace: 'AWS/ElastiCache',
        metricName: 'CurrConnections',
        dimensions: { CacheClusterId: cacheClusterId },
        statistic: 'avg',
      }),
      threshold,
      period,
      evaluationPeriods: 1,
      alarmDescription: `The connection constant in ${period.toMinutes()} minute(s) is greater than ${threshold}.`,
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      treatMissingData: cloudwatch.TreatMissingData.NOT_BREACHING,
    });
  }
}

src/index.ts
import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as sns from '@aws-cdk/aws-sns';
import * as sns_sub from '@aws-cdk/aws-sns-subscriptions';
import * as actions from '@aws-cdk/aws-cloudwatch-actions';
import { CpuMonitor } from './cpu-monitor';
import { SwapUsageMomiton } from './swap-usage-monitor';
import { EvictionsMomiton } from './evictions-monitor';
import { CurrConnectionsMomiton } from './current-connections-monitor';
import { FreeableMemoryMomiton } from './freeable-memory-monitor';

/**
 * Elasticache auto monitor set up base props.
 */
interface ISetUpProps {
  /**
   * Default elasticache node type.
   * It is strongly recommended to set according to the actual value.
   *
   * @default NodeType.M5_LARGE
   */
  readonly nodeType?: NodeType;

  /**
   * The period over which the specified statistic is applied.
   *
   * @default Duration.minutes(1)
   */
  readonly currConnectionsPeriod?: cdk.Duration;

  /**
   * CurrConnections threshold
   * 
   * Every 1 vcup is equal to 50 connections
   * 
   * @default 100
   */
  readonly currConnectionsThreshold?: number;
}

/**
 * Elasticache auto monitor set up with labmda props.
 */
export interface ISetUpWithLambdaProps extends ISetUpProps {
  readonly lambda: lambda.Function,
}

export interface ISetUpWithSlackProps extends ISetUpProps {

  /**
   * Go to this(https://slack.com/apps/A0F7XDUAZ-incoming-webhooks) link to apply for webhook 
   */
  readonly webHookUrl: string;

  /**
   * Setting channel.
   * What is a channel: (https://slack.com/intl/en-cn/help/articles/360017938993-What-is-a-channel)
   *
   * @default cloudwatch-alarm
   */
  readonly channel?: string;

  /**
   * Setting Slack bot name.
   * 
   * @default Webhookbot
   */
  readonly username?: string;

  /**
   * Emoji for bot.
   * 
   * @default :scream:
   */
  readonly iconEmoji?: string;
}

export interface ISetUpWithSmsProps extends ISetUpProps {
  /**
   * Include country code and phone number, for example: +15551231234
   */
  readonly phones: string[];
}

/**
 * ElasticacheAutoMonitor allows you to send email, sms, slack, or trigger aws sns topic when an alarm occurs.
 * You will get the following monitoring:
 *  1. Cpu Monitor: Should be less than 90%. (See below reference)
 *  2. SwapUsage Monitor: Should be less than 50M.
 *  3. Evictions Monitor: Should not have evictions value.
 *  4. CurrConnections Monitor: According to your business needs, default every 1 vcup is equal to 200 connections.
 *  5. FreeableMemory Monitor: Not less than 10%
 * 
 * Reference: https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheMetrics.WhichShouldIMonitor.html
 */
export class ElasticacheAutoMonitor extends cdk.Construct {

  public static setUpWithLambda(scope: cdk.Construct, cacheClusterId: string, props: ISetUpWithLambdaProps) {
    const topic = new sns.Topic(scope, 'alarm-' + cacheClusterId + '-lambda');
    topic.addSubscription(new sns_sub.LambdaSubscription(props.lambda));
    ElasticacheAutoMonitor.setup(scope, cacheClusterId, topic, props);
  }

  public static setUpWithSlack(scope: cdk.Construct, cacheClusterId: string, props: ISetUpWithSlackProps) {

    let channel = props.channel || 'cloudwatch-alarm';
    let username = props.username || 'Webhookbot';
    let icon_emoji = props.iconEmoji || ':scream:';

    const fn = new lambda.Function(scope, 'alarm-' + cacheClusterId + '-to-slack', {
      handler: 'index.handler',
      runtime: lambda.Runtime.PYTHON_3_8,
      code: lambda.Code.fromAsset(path.join(__dirname, '../asset/lambda')),
      environment: {
        SLACK_WEBHOOK_URL: props.webHookUrl,
        CHANNEL: channel,
        USERNAME: username,
        ICON_EMOJI: icon_emoji,
      },
    });

    const topic = new sns.Topic(scope, 'alarm-' + cacheClusterId + '-slack');
    topic.addSubscription(new sns_sub.LambdaSubscription(fn));

    ElasticacheAutoMonitor.setup(scope, cacheClusterId, topic, props);
  }

  public static setUpWithSms(scope: cdk.Construct, cacheClusterId: string, props: ISetUpWithSmsProps) {
    const topic = new sns.Topic(scope, 'alarm-' + cacheClusterId + '-with-sms-topic');
    for (let phone in props.phones) {
      topic.addSubscription(new sns_sub.SmsSubscription(phone));
    }
    ElasticacheAutoMonitor.setup(scope, cacheClusterId, topic, props);
  }

  private static setup(scope: cdk.Construct, cacheClusterId: string, topic: sns.Topic, props: ISetUpProps) {

    // bind default and props
    const nodeType = props.nodeType || NodeType.DEFAULT;
    const _props = {
      nodeType,
      currConnectionsPeriod: cdk.Duration.minutes(1),
      currConnectionsThreshold: nodeType.vcupCount * 200,
      ...props,
    }

    CpuMonitor.setup(scope, cacheClusterId, _props.nodeType).addAlarmAction(new actions.SnsAction(topic));
    SwapUsageMomiton.setup(scope, cacheClusterId).addAlarmAction(new actions.SnsAction(topic));
    EvictionsMomiton.setup(scope, cacheClusterId).addAlarmAction(new actions.SnsAction(topic));
    CurrConnectionsMomiton.setup(scope, cacheClusterId, _props.currConnectionsPeriod, _props.currConnectionsThreshold).addAlarmAction(new actions.SnsAction(topic));
    FreeableMemoryMomiton.setup(scope, cacheClusterId, _props.nodeType).addAlarmAction(new actions.SnsAction(topic));
  }

}

export class NodeType {

  readonly name: string;
  readonly vcupCount: number;
  readonly memory: number;

  private constructor(name: string, vcupCount: number, memory: number) {
    this.name = name;
    this.vcupCount = vcupCount;
    this.memory = memory;
  }

  public static readonly T2_MIRCO = new NodeType('cache.t2.mirco', 1, 0.555 * 1024 * 1024 * 1024);
  public static readonly T2_SMALL = new NodeType('cache.t2.small', 1, 1.55 * 1024 * 1024 * 1024);
  public static readonly T2_MEDIUM = new NodeType('cache.t2.medium', 2, 3.22 * 1024 * 1024 * 1024);
  public static readonly T3_MICRO = new NodeType('cache.t3.micro', 2, 0.5 * 1024 * 1024 * 1024);
  public static readonly T3_SMALL = new NodeType('cache.t3.small', 2, 1.37 * 1024 * 1024 * 1024);
  public static readonly T3_MEDIUM = new NodeType('cache.t3.medium', 2, 3.09 * 1024 * 1024 * 1024);
  public static readonly M4_LARGE = new NodeType('cache.m4.large', 2, 6.42 * 1024 * 1024 * 1024);
  public static readonly M4_XLARGE = new NodeType('cache.m4.xlarge', 4, 14.28 * 1024 * 1024 * 1024);
  public static readonly M4_2XLARGE = new NodeType('cache.m4.2xlarge', 8, 29.70 * 1024 * 1024 * 1024);
  public static readonly M4_4XLARGE = new NodeType('cache.m4.4xlarge', 16, 60.78 * 1024 * 1024 * 1024);
  public static readonly M4_10XLARGE = new NodeType('cache.m4.10xlarge', 40, 154.64 * 1024 * 1024 * 1024);
  public static readonly M5_LARGE = new NodeType('cache.m5.large', 2, 6.38 * 1024 * 1024 * 1024);
  public static readonly M5_XLARGE = new NodeType('cache.m5.xlarge', 4, 12.93 * 1024 * 1024 * 1024);
  public static readonly M5_2XLARGE = new NodeType('cache.m5.2xlarge', 8, 26.04 * 1024 * 1024 * 1024);
  public static readonly M5_4XLARGE = new NodeType('cache.m5.4xlarge', 16, 52.26 * 1024 * 1024 * 1024);
  public static readonly M5_12XLARGE = new NodeType('cache.m5.12xlarge', 48, 157.12 * 1024 * 1024 * 1024);
  public static readonly M5_24XLARGE = new NodeType('cache.m5.12xlarge', 96, 314.32 * 1024 * 1024 * 1024);
  public static readonly R4_LARGE = new NodeType('cache.r4.large', 2, 12.3 * 1024 * 1024 * 1024);
  public static readonly R4_XLARGE = new NodeType('cache.r4.xlarge', 4, 25.05 * 1024 * 1024 * 1024);
  public static readonly R4_2XLARGE = new NodeType('cache.r4.2xlarge', 8, 50.47 * 1024 * 1024 * 1024);
  public static readonly R4_4XLARGE = new NodeType('cache.r4.4xlarge', 16, 101.38 * 1024 * 1024 * 1024);
  public static readonly R4_8XLARGE = new NodeType('cache.r4.8xlarge', 32, 203.26 * 1024 * 1024 * 1024);
  public static readonly R4_16XLARGE = new NodeType('cache.r4.16xlarge', 64, 407 * 1024 * 1024 * 1024);
  public static readonly R5_LARGE = new NodeType('cache.r5.large', 2, 13.07 * 1024 * 1024 * 1024);
  public static readonly R5_XLARGE = new NodeType('cache.r5.xlarge', 4, 26.32 * 1024 * 1024 * 1024);
  public static readonly R5_2XLARGE = new NodeType('cache.r5.2xlarge', 8, 52.82 * 1024 * 1024 * 1024);
  public static readonly R5_4XLARGE = new NodeType('cache.r5.4xlarge', 16, 105.81 * 1024 * 1024 * 1024);
  public static readonly R5_12XLARGE = new NodeType('cache.r5.12xlarge', 48, 317.77 * 1024 * 1024 * 1024);
  public static readonly R5_24LARGE = new NodeType('cache.r5.12xlarge', 96, 635.61 * 1024 * 1024 * 1024);

  public static readonly DEFAULT = NodeType.M5_LARGE;
}
